pkg/envoy/lds/builder.go

```diff
@@ -0,0 +1,599 @@
+	lb.listenerFilters = append(lb.listenerFilters,
+		&xds_listener.ListenerFilter{
+			// To inspect TLS metadata, such as the transport protocol and SNI
+			Name: envoy.TLSInspectorFilterName,
+			ConfigType: &xds_listener.ListenerFilter_TypedConfig{
+				TypedConfig: &any.Any{
+					TypeUrl: envoy.TLSInspectorFilterTypeURL,
+				},
+			},
+		},
+		&xds_listener.ListenerFilter{
+			// The OriginalDestination ListenerFilter is used to restore the original destination address
+			// as opposed to the listener's address (due to iptables redirection).
+			// This enables  filter chain matching on the original destination address (ip, port).
+			Name: envoy.OriginalDstFilterName,
+			ConfigType: &xds_listener.ListenerFilter_TypedConfig{
+				TypedConfig: &any.Any{
+					TypeUrl: envoy.OriginalDstFilterTypeURL,
+				},
+			},
+		},
+	)
+	return lb
+}
+
+func (lb *listenerBuilder) TracingEndpoint(endpoint string) *listenerBuilder {
+	lb.httpTracingEndpoint = endpoint
+	return lb
+}
+
+func (lb *listenerBuilder) ExtAuthzConfig(config *auth.ExtAuthConfig) *listenerBuilder {
+	lb.extAuthzConfig = config
+	return lb
+}
+
+func (lb *listenerBuilder) WASMStatsHeaders(headers map[string]string) *listenerBuilder {
+	lb.wasmStatsHeaders = headers
+	return lb
+}
+
+func (lb *listenerBuilder) ActiveHealthCheck(enable bool) *listenerBuilder {
+	lb.activeHealthCheck = enable
+	return lb
+}
+
+func (lb *listenerBuilder) TrafficTargets(t []trafficpolicy.TrafficTargetWithRoutes) *listenerBuilder {
+	lb.trafficTargets = t
+	return lb
+}
+
+func (lb *listenerBuilder) TrustDomain(trustDomain string) *listenerBuilder {
+	lb.trustDomain = trustDomain
+	return lb
+}
+
+func (lb *listenerBuilder) SidecarSpec(sidecarSpec configv1alpha2.SidecarSpec) *listenerBuilder {
+	lb.sidecarSpec = sidecarSpec
+	return lb
+}
+
+func (lb *listenerBuilder) Build() (*xds_listener.Listener, error) {
+	l := &xds_listener.Listener{
+		Name:             lb.name,
+		Address:          lb.address,
+		TrafficDirection: lb.trafficDirection,
+		ListenerFilters:  lb.listenerFilters,
+		AccessLog:        envoy.GetAccessLog(),
+	}
+
+	var outboundTrafficMatches []*trafficpolicy.TrafficMatch // used to configure FilterDisabled match predicate
+	switch l.TrafficDirection {
+	case xds_core.TrafficDirection_OUTBOUND:
+		l.FilterChains = append(l.FilterChains, lb.buildOutboundFilterChains()...)
+		if lb.outboundMeshTrafficPolicy != nil {
+			outboundTrafficMatches = append(outboundTrafficMatches, lb.outboundMeshTrafficPolicy.TrafficMatches...)
+		}
+
+	case xds_core.TrafficDirection_INBOUND:
+		l.FilterChains = append(l.FilterChains, lb.buildInboundFilterChains()...)
+
+	default:
+		return nil, fmt.Errorf("listener %s: unsupported traffic direction %s", l.Name, l.TrafficDirection)
+	}
+
+	if lb.permissiveEgress {
+		// Enable permissive (global) egress to unknown destinations
+		egressFilterChain, err := getDefaultPassthroughFilterChain()
```

Comment:
That's a good suggestion. Some of the protos that are dynamically generated need to be verified if the marshalling is correct, but agree about the static configs. I'll leave this as is for now since unit tests would detect static config marshalling issues at present.